pyvista/utilities/features.py

```diff
@@ -56,6 +56,15 @@ def voxelize(mesh, density=None, check_surface=True):
     if isinstance(density, (list, set, tuple)):
         density_x, density_y, density_z = density
 
+    # check and pre-process input mesh
+    surface = mesh.extract_geometry()  # filter preserves topology
```

Comment:
I replaced `extract_surface()` with `extract_geometry()`. According to https://vtk.org/doc/nightly/html/classvtkGeometryFilter.html#details vs. https://vtk.org/doc/nightly/html/classvtkDataSetSurfaceFilter.html the trade-offs are not obvious in modern VTK. Both filters delegate to the other for linear/non-linear cells, so either case is probably optimal with vanilla arguments.

But crucially `extract_geometry()` preserves the topology, so we end up with manifold meshes when applying it to volumes. For the equivalent behaviour we'd need `extract_surface().clean()`. I suspect that would be overall slower, and the docs mention that `vtkDataSetSurfacefilter` used by `extract_surface()` can also use a lot of memory. If there are potential edge cases where this change might matter, I'd gladly add test cases.

Reply:
`clean` is slow primarily because it checks for duplicates (there's some intermediate data structure it creates), and I'm fine with using `extract_geometry`.

As for the edge cases, when those come up we can check for them. For the time being, this PR is created because of an edge case in the first place, so I'm fine with going down the `extract_geometry` route.